segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/chunklayer/GarbageCollector.java

```diff
@@ -0,0 +1,182 @@
+/**
+ * Copyright (c) Dell Inc., or its subsidiaries. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ */
+package io.pravega.segmentstore.storage.chunklayer;
+
+import io.pravega.common.Exceptions;
+import io.pravega.common.concurrent.Futures;
+import io.pravega.segmentstore.storage.metadata.ChunkMetadata;
+import lombok.Getter;
+import lombok.extern.slf4j.Slf4j;
+import lombok.val;
+
+import java.time.Duration;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+/**
+ * Implements simple garbage collector for cleaning up the deleted chunks.
+ */
+@Slf4j
+public class GarbageCollector implements AutoCloseable {
+    /**
+     * Set of garbage chunks.
+     */
+    @Getter
+    private final ConcurrentHashMap<String, Long> garbageChunks = new ConcurrentHashMap<>();
+
+    private final ChunkedSegmentStorage chunkedSegmentStorage;
+
+    private final ChunkedSegmentStorageConfig config;
+
+    private final AtomicBoolean closed = new AtomicBoolean();
+
+    private final AtomicBoolean suspended = new AtomicBoolean();
+
+    private final long delayInMillis;
+
+    private CompletableFuture<Void> loopFuture;
+
+    /**
+     * Constructs a new instance.
+     *
+     * @param chunkedSegmentStorage Instance of {@link ChunkedSegmentStorage}.
+     * @param config Configuration to use.
+     */
+    public GarbageCollector(ChunkedSegmentStorage chunkedSegmentStorage, ChunkedSegmentStorageConfig config) {
+        this.chunkedSegmentStorage = chunkedSegmentStorage;
+        this.config = config;
+        this.delayInMillis = Duration.ofSeconds(config.getGarbageCollectionFrequencyInSeconds()).toMillis();
+    }
+
+    /**
+     * Initializes this instance.
+     */
+    public void initialize() {
+        loopFuture = run();
+    }
+
+    public CompletableFuture<Void> run() {
+        return Futures.loop(
+                () -> !closed.get(),
+                () -> CompletableFuture.completedFuture(null)
+                        .thenComposeAsync(v -> deleteGarbage(true, config.getGarbageCollectionConcurrency()),
+                                chunkedSegmentStorage.getExecutor())
+                        .handleAsync((v, ex) -> {
+                            if (null != ex) {
+                                log.error("{} Error during run {}.", chunkedSegmentStorage.getLogPrefix(), ex);
```

Comment:
This will double-log the exception. Remove the second `{}`.

Reply:
ok.